taichi/system/timeline.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include <vector>
+#include <mutex>
+
+#include "taichi/common/core.h"
+#include "taichi/system/timer.h"
+
+TI_NAMESPACE_BEGIN
+
+struct TimelineEvent {
+  std::string name;
+  bool begin;
+  float64 time;
+  std::string tid;
+
+  std::string to_json();
+};
+
+class Timeline {
+ public:
+  Timeline();
+
+  ~Timeline();
+
+  static Timeline &get_this_thread_instance();
+
+  void set_name(const std::string &tid) {
+    tid_ = tid;
+  }
+
+  std::string get_name() {
+    return tid_;
+  }
+
+  void clear();
+
+  void insert_event(const TimelineEvent &e);
+
+  std::vector<TimelineEvent> fetch_events();
+
+  class Guard {
+   public:
+    Guard(const std::string &name);
+
+    ~Guard();
+
+   private:
+    std::string name_;
+  };
+
+ private:
+  std::string tid_;
+  std::mutex mut_;
+  std::vector<TimelineEvent> events_;
+};
+
+// A timeline system for multi-threaded applications
+class Timelines {
+ public:
+  static Timelines &get_instance();
+
+  void insert_events(const std::vector<TimelineEvent> &events,
```

Comment:
(Funny that Github sent my comment to `/dev/null`...)

Usually this can be solved by separating the function into two:

```cpp
 public:
  void insert_events(const std::vector<TimelineEvent> &events);  // |mut_| must NOT be held

 private:
  void insert_events_locked(const std::vector<TimelineEvent> &events);  // |mut_| MUST be held
```

In the future we can also annotate the locking behavior with https://clang.llvm.org/docs/ThreadSafetyAnalysis.html  :)

Reply:
Cool - let me fix this. Thanks for the tip :-)